### ISSUE TITLE
"Error: Callback was already called" when using "near" filter

### ISSUE THREAD

**Issue by bachirelkhoury** [strongloop/loopback-connector-mongodb#285]
Given this simple geolocation query, when I include the "near" query I'm getting this error:
```
/node_modules/mongodb/lib/utils.js:98
    process.nextTick(function() { throw err; });
                                  ^

Error: Callback was already called.
    at /node_modules/async/dist/async.js:844:36
    at /node_modules/async/dist/async.js:3676:17
    at /node_modules/async/dist/async.js:339:31
    at /node_modules/strong-remoting/lib/remote-objects.js:694:7
    at execStack (/node_modules/strong-remoting/lib/remote-objects.js:490:9)
    at RemoteObjects.execHooks (/node_modules/strong-remoting/lib/remote-objects.js:497:10)
    at interceptInvocationErrors (/node_modules/strong-remoting/lib/remote-objects.js:693:10)
    at /node_modules/loopback-phase/node_modules/async/lib/async.js:148:21
    at /node_modules/loopback-phase/node_modules/async/lib/async.js:148:21
    at /node_modules/loopback-phase/node_modules/async/lib/async.js:148:21
```

Can't find the root cause of this.

This code works:

```
user.find({}, function(err, users) {
        if (err) {
          return cb(err);
        }
        return cb(null, users);
    });
  };
```
This one below throws the error. I've placed a breakpoint in the callback, and it does get called twice internally. Couldn't trace it too far so not sure what's firing this.

```
user.find({
      where: {
        location: {near: here, maxDistance: distance, unit: 'meters'}
      }
    }, function(err, users) {
        if (err) {
          return cb(err);
        }
        return cb(null, users);
    });
  };
```

**Comment by Amir-61** [comment]
Possibly duplicate of https://github.com/strongloop/loopback-datasource-juggler/issues/1102

**Comment by rptally** [comment]
I had another version of the server code running on a different machine that wasn't showing the problem.  The only dependency I changed to correct the issue was the loopback-datasource-juggler version back from 51 to 50 to 49.  The loopback-connector-mongodb performs the query correctly with loopback-datasource-juggler 2.49.0.

This is the dependencies section from the project package.json  

    "dependencies": {
      "compression": "^1.6.2",
      "cors": "^2.8.1",
      "helmet": "^1.3.0",
      "loopback": "^2.34.1",
      "loopback-boot": "^2.22.0",
      "loopback-component-explorer": "^2.6.0",
      "loopback-component-storage": "^1.9.1",
      "loopback-connector-mongodb": "^1.15.2",
      "loopback-connector-sparkpost": "0.0.2",
      "loopback-datasource-juggler": "~2.49.0",
      "serve-favicon": "^2.3.0",
      "strong-error-handler": "^1.1.0",
      "twilio": "^2.11.0"
  },

**Comment by Amir-61** [comment]
@rptally @bachirelkhoury it seems this is a regression from a recent work. Could @rptally or @bachirelkhoury  please provide a sample repo using our loopback-sandbox so I can reproduce it on my local machine and then label it as a bug and escalate it?

I suspect this is a regression from a recent community [PR#259](https://github.com/strongloop/loopback-connector-mongodb/pull/259); @superkhau since you are reviewed that PR, you may have a better insight.

**Comment by rptally** [comment]
It's very easy to duplicate.  Create a loopback app; add a mongo datasource; create a model (Store) with a geopoint field called loc; start the server.   Then use the browser to run the query:

    http://localhost:3000/api/stores?filter[where][loc][near]=32.0,-92.0

You immediately get the "Error: Callback was already called." error.

I've attached the project as requested.  Please let me know what else you need.

[fred.zip](https://github.com/strongloop/loopback-connector-mongodb/files/483562/fred.zip)

**Comment by superkhau** [comment]
@Amir-61 Thanks for verifying the bug. Let's escalate/TOB/major it.

**Comment by superkhau** [comment]
@rptally What node version are you on?

**Comment by superkhau** [comment]
As for the regression, that part is not for sure yet. @loay Is looking into a similiar failure to confirm or see a related fix will fix this too.

**Comment by rptally** [comment]
@superkhau v4.5.0

fred@dal-api-1:~# node -v
v4.5.0

**Comment by jainisenough** [comment]
issue exist with too
node: 5.11.1
"loopback-connector-mongodb": "^1.15.1"

**Comment by superkhau** [comment]
@Amir-61 has been able to reproduce the error and is working on a fix. It is a regression introduced at https://github.com/strongloop/loopback-datasource-juggler/pull/1061.

If @jainisenough @rptally @bachirelkhoury want to take a crack at it, ping me and we can get the fix landed sooner.

**Comment by geodeveloper** [comment]
I can reproduce this bug too, by using the **near** filter with datasource MongoDB.

**Comment by jwebcat** [comment]
This bug exists with postgres connector as well.

**Comment by Zvika88** [comment]
It happens with either Mongo or Postgres. Distance query causes the server to crash with the "Callback was already called" error. Wow, that is a show-stopper issue. @superkhau , would you mind sharing if there is an ETA on a fix for that?

**Comment by Amir-61** [comment]
Hey guys we have moved it to top of backlog, and we will take care of it soon. Unfortunately there was a regression from a recent work and there was no test coverage to know that the changes break this geo query as a result of lacking test coverage for geo query. Sorry for inconvenience. For now you may use `loopback-datasource-juggler@2.49.0` to get rid of this bug. We will take care of it soon.

Thanks!

**Comment by superkhau** [comment]
@Zvika88 @Amir-61 is working on the issue ATM and has better insight into the ETA. It will definitely be pulled into planning for next sprint (ie. next Tuesday). The workaround for now as mentioned by @Amir-61 is to use juggler@2.49.0 until the fix gets landed.

**Comment by nammaianh** [comment]
Why 2.49.0 works, but 2.51.0 does not ❓

**Comment by Amir-61** [comment]
@rptally @bachirelkhoury @jainisenough @geodeveloper @jwebcat @Zvika88

https://github.com/strongloop/loopback-datasource-juggler/pull/1108 fixes the bug reported here. I tried https://github.com/strongloop/loopback-datasource-juggler/pull/1108 and it seems the bug reported here gets resolved; before landing the patch I also appreciate if one of you please verify that the bug you reported here is fixed by https://github.com/strongloop/loopback-datasource-juggler/pull/1108 for juggler@2.x.

All you need to do is to:
- Clone `loopback-datasource-juggler`
- Go to feature branch which fixes the issue reported here `git checkout fix_geo_2.x`
- Once you are in the cloned repo try `npm link`
- Go to your project and try `npm link loopback-datasource-juggler`
- Now please verify the bug you used to have is now fixed in your project with whatever connector you were using.

Thanks a lot!

**Comment by Amir-61** [comment]
To make it simpler, I uploaded [loopback-geopoint-example](https://github.com/Amir-61/loopback-geopoint-example), which has a simple geopoint query with `near` filter using `loopback-connector-mongodb` as a datasource for backing system. If you run it as is with `"loopback-datasource-juggler": "^2.51.0"`, it fails. However if you do the steps mentioned in the previous comment using `npm link`, it works fine. So I'm going to land the PR after getting final LGTM from @bajtos for https://github.com/strongloop/loopback-datasource-juggler/pull/1108

**Comment by bajtos** [comment]
> All you need to do is to (...)

FWIW, npm can install directly from github. `npm install strongloop/loopback-datasource-juggler#fix_geo_2.x` should work equally well.

**Comment by Amir-61** [comment]
https://github.com/strongloop/loopback-datasource-juggler/pull/1108 landed and released `loopback-datasource-juggler@2.53.0`

**Comment by Amir-61** [comment]
https://github.com/strongloop/loopback-datasource-juggler/pull/1108 landed and released loopback-datasource-juggler@2.53.0.

I will be closing this issue.

Thanks!

**Comment by grs4** [comment]
Hello Amit-61

I am stuck with the same error.  

ref issue# 3667
https://github.com/strongloop/loopback/issues/3667

I am using loopback 3.x & version of juggler and postgresql connector as below

loopback-datasource-juggler@^3.9.3 
loopback-connector-postgresql: "^3.1.0"

Please advise if there is any patch available.
